From 460d2d999990437170aeed22cf2d143608b3218f Mon Sep 17 00:00:00 2001
From: Paul Berry <stereotype441@gmail.com>
Date: Wed, 14 Nov 2012 15:19:07 -0800
Subject: [PATCH 29/30] i965: Enable offloading GL to a separate thread through
 a driconf option.

We don't know yet when it will make sense to enable by default, so make it
opt-in for the moment.
---
 src/mesa/drivers/dri/i915/intel_screen.c | 4 ++++
 src/mesa/drivers/dri/i965/brw_context.c  | 7 +++++++
 2 files changed, 11 insertions(+)

diff --git a/src/mesa/drivers/dri/i915/intel_screen.c b/src/mesa/drivers/dri/i915/intel_screen.c
index 9af06275e0..3b5b6a8fa5 100644
--- a/src/mesa/drivers/dri/i915/intel_screen.c
+++ b/src/mesa/drivers/dri/i915/intel_screen.c
@@ -63,6 +63,10 @@ DRI_CONF_BEGIN
 	 DRI_CONF_DESC(en, "Enable early Z in classic mode (unstable, 945-only).")
       DRI_CONF_OPT_END
 
+      DRI_CONF_OPT_BEGIN(glthread, bool, false)
+	 DRI_CONF_DESC(en, "Enable offloading GL driver work to a separate thread")
+      DRI_CONF_OPT_END
+
    DRI_CONF_SECTION_END
    DRI_CONF_SECTION_QUALITY
       DRI_CONF_FORCE_S3TC_ENABLE("false")
diff --git a/src/mesa/drivers/dri/i965/brw_context.c b/src/mesa/drivers/dri/i965/brw_context.c
index 3705a9140c..025a82b419 100644
--- a/src/mesa/drivers/dri/i965/brw_context.c
+++ b/src/mesa/drivers/dri/i965/brw_context.c
@@ -43,6 +43,7 @@
 #include "main/vtxfmt.h"
 #include "main/texobj.h"
 #include "main/framebuffer.h"
+#include "main/glthread.h"
 
 #include "vbo/vbo_context.h"
 
@@ -1154,6 +1155,12 @@ brwCreateContext(gl_api api,
    _mesa_initialize_dispatch_tables(ctx);
    _mesa_initialize_vbo_vtxfmt(ctx);
 
+   if (driContextPriv->driScreenPriv->dri2.backgroundCallable &&
+       driQueryOptionb(&brw->optionCache, "glthread")) {
+      /* Loader supports multithreading, and so do we. */
+      _mesa_glthread_init(ctx);
+   }
+
    vbo_use_buffer_objects(ctx);
    vbo_always_unmap_buffers(ctx);
 
-- 
2.11.0

